### ISSUE TITLE
Rare exposure check

### ISSUE THREAD

**Issue by fabioalbani** [immuni-app/immuni-app-android#233]
# Issue content
I've noticed that in ten days Google covid exposure notifications, with Immuni, checked only 3 times for potential contacts with positives. Is it correct? At the moment I'm a remote worker so my phone can connect only with my wife and my tablet, or some customer in shops, but I can't imagine that checks are so rare.. I would imagine checks like Facebook notifications... Minutes... I have a Huawei p8lite 2017.

**Comment by fabioalbani** [comment]
On my tablet, less used but a Lenovo Tablet, is more frequent anyway, like a check each 2 days.

**Comment by filippogasbarro** [comment]
Similar thing to me:
By comparing with another person, I noticed that Huawei does less checks on the last covid-19 exposure check. For example, on Huawei P20lite the last exposure control is 13 June.  While in the Samsung A5 2017 it has performed 4 or 5 more checks in this period of time until today.
 Is there any problem with Huawei?  The immuni app it was opened at least once every two days in both smartphone

**Comment by fabioalbani** [comment]
The same for me... Moreover, I've opened Immuni on my tablet once, while on my smartphone Huawei I'm opening it each day. I've tried to update the app, delete covid notifications, restart it, to see if the check would be performed.. But the last one is of the 15th of June... If I would be noticed of a contact probably my quarantine would be expired..

**Comment by filippogasbarro** [comment]
Exactly "If I would be noticed of a contact probably my quarantine would be expired.."
My last exposure control is 13 of June on Huawei P20 lite. Today I rebooted the smartphone and the last check is today. 🤔
In last two weeks, I opened at least one time every two days (as I said in a post before) and also I switch off the battery saving for the app Google Play services and Immuni

**Comment by fabioalbani** [comment]
After a shutdown also my smartphone checked for notifications... Is it possible that checks are done at startup? I shut down the tablet while my phone is always on...

**Comment by lmasellis** [comment]
Right now, the latest exposure key export file (no. 13) served by the backend server has "end_timestamp: 1593025200", which translates to Wed Jun 24 21:00:00 CEST 2020.

But on Google exposure notification settings page latest exposure check reads June 17th at 14:45.

Either I do not understand the meaning of the date on the settings page, or the Immuni app is not feeding the keys to exposure notification service.

Not that it would be a great deal of trouble, since the number of keys served by the backend seems VERY VERY scarce.

Only 13 files were served so far, and each of them contains a very small number of keys. Looking more carefully at file no. 13, for example, one can find out that all the keys in each single file form a complete sequence of rolling_start_interval_number with an increment of 144 between neighbouring values, which realistically means all keys in each file come from the same positive-reporting user. Other served files are similar. So one must conclude that only about 13 confirmed positive users have been notified by the backend so far. A little bit disappointing, if my understanding is right...

**Comment by riccardocossu** [comment]
I have a similar problem; it seem to check ony once after reboot, and it does it multiple times in a row (same timestamp). Thus I only have 9 checks so far, 6 on June 24th at 9.01 and 3 on june 28th at 21.53 (after a reboot).
All checks report 0 as "numer of keys", so I guess it didn't download anything in fact

**Comment by fabioalbani** [comment]
My Google play services has been updated or set remotely, now I see a kind of a red sun light at the left of the exposure check entry in Google services options. Now the exposure check are on a daily basis so the problems seems fixed.

**Comment by riccardocossu** [comment]
@fabioalbani not sure if I have the same Play Services version you have, nor if this matters; I still have the problem

**Comment by filippogasbarro** [comment]
Now it seems fixed (I think by an update of play services). In these last days the app have done two check without rebooting the phone. The last check (yesterday 23:41) was done also by not opening Immuni app 🙂

**Comment by filippogasbarro** [comment]
@riccardocossu is the problem resolved now? It's about one week that immuni do the automatic check without rebooting the phone

**Comment by riccardocossu** [comment]
@filippogasbarro no, not solved, but I can give more detail on what's happening: it only does the exposures check after a "fresh start", that is the first time the app is started either after reboot or after force killing the app from the OS.
And it does a numbers of checks equal to the number of days that have passed since the latest check; for example if it last checked 3 days ago, it does 3 checks in a row, if it last checked yesterday it only does one

**Comment by filippogasbarro** [comment]
Tried to update to latest version of Google Play services?
Use this link in your smartphone to check if is it updated:
[](https://play.google.com/store/apps/details?id=com.google.android.gms&hl=it)

**Comment by riccardocossu** [comment]
It got worse, meaning that killing the app doesn't work anymore; I rebooted and there has been an improvement: even though I skipped a few days, after reboot the file has been download just once, even though still empty (0 keys).
I will test further in the next days

**Comment by andreapede** [comment]
I have the same behaviour described before @riccardocossu . Update only after a fresh start and a line created for the amount of days since the previous check. In any case inside always 0 keycount and 0 matchescount (luckly).

Google Play Service 202414037. Xiaomi Mi8.

**Comment by Teudimundo** [comment]
Huwaei P20 and Huwaei P20 Lite: the first last 14 days 8 checks, the 8 checks have been done 13 days ago and they all have the same time (20:19). The latter report a last check more than 16 days ago.

**Comment by andreapede** [comment]
Just noticed that after
Deleting all the Google covid data 
Uninstalling and reinstalling the latest immuni version
The keys started to be downloaded.

**Comment by TheShrike** [comment]
Same problem on a pixel phone, 0 checks in the last 14 days, and always 0 keycount and 0 matches.

**Comment by filippogasbarro** [comment]
@andreapede so there aren't any other way by non deleting all 15 days keys collected?

**Comment by andreapede** [comment]
> @andreapede so there aren't any other way by non deleting all 15 days keys collected?

I couldn't find any other workaround. Just removing the app and cleaning the cache doesn't solve the issue.

**Comment by lmasellis** [comment]
The team should provide an official position on this. If it's only a GUI issue, then it would be reassuring. If, instead, all early users are relying on an app that will not work unless cleared and reinstalled, they should be notified ASAP (though, I must admit, this would be a serious blow for the reputation of the system).

**Comment by andreapede** [comment]
For sake of completeness: I installed the app since the first release going through the various update. Couple of time I uninstalled and reinstalled again to verify the change in the received keys (without any result). This is the first time I cleaned also the covid framework. I also installed the 1.3 version on another never used phone and it seems working without problem. So it really seems that some issue occurred only for the really earlier installation.

**Comment by lmasellis** [comment]
Update: I just noticed that today, for the first time, a check with non-zero number of keys was reported. 

I did not clear the keys, I did not reinstall the app or willingly did anything else that could lead to this change.

But I noticed subtle changes in the appearance of the Google Exposure Notification settings menu, even though the Google Play Services version is still the same (20.24.14).

![Screenshot_20200724-135808~2](https://user-images.githubusercontent.com/65446168/88391530-a9868d00-cdba-11ea-90f9-b1a9171012c1.png)

Nonetheless, it would be interesting to know whether the keys were actually checked before today or not.

**Comment by riccardocossu** [comment]
None of the workarounds worked for me; the app does not deliver its basic feature, and it doesn't seem to be related to xiaomi phones as I first thought. Definitely the company must address this, at this point the app is useless for me

**Comment by filippogasbarro** [comment]
Without doing anything. Today 17 keys on my Huawei 😃
![Screenshot_20200724_175355_com google android gms](https://user-images.githubusercontent.com/15571375/88410602-e06a9c00-cdd6-11ea-900f-d57b131aff32.jpg)

**Comment by Kariamos** [comment]
@filippogasbarro is it still working okay?

**Comment by lmasellis** [comment]
> @filippogasbarro is it still working okay?

For me, no.

It seemed to work correctly for a couple of days (frequent checks reported and non-zero key numbers).

Now it's stuck with last check on July 24th.

**Comment by andreapede** [comment]
> > @filippogasbarro is it still working okay?
> 
> For me, no.
> 
> It seemed to work correctly for a couple of days (frequent checks reported and non-zero key numbers).
> 
> Now it's stuck with last check on July 24th.

The cleaning of data framework and app reinstallation only solved the issue to receive non-zero key any time. But as usual the data exchange happens only when the app is started after a reboot or a forced stop (so at the real start).

**Comment by filippogasbarro** [comment]
@Kariamos yes it work. Today at 18:00 10 keys

**Comment by Kariamos** [comment]
For me too worked yesterday, maybe for the first time... will the check be at the same time everyday or random?

**Comment by riccardocossu** [comment]
same for me, I reinstalled the app a couple of days ago and it seems to work as intended; it checks for exposure without reboots and it usually reports a non-zero number of keys. It sometimes checks once a day, sometimes more than once.

**Comment by andreapede** [comment]
After couple of weeks in which the app was working as designed, now starts again to receive data only after a reboot. To.a first look it seems that nor framework neither app version changed.
I'm honestly confused by this erratic behavior

**Comment by riccardocossu** [comment]
sadly I also have this defect again; it worked ok for a couple of days then it stopped like before. May this be related to the fact that I often leave the BT off? I turn it off when I am at home alone, since it would be a waste of battery for no gain...

**Comment by riccardocossu** [comment]
@andreapede I found out I was actually wrong; the service kept on checking exposures (I have no missing day, most of them have 2 or 3 checks each). It's just that the list in Google options is rendered in no meaningful order: it's not date, it's not number of keys... maybe alphabetical order of the hash? But this is definitely not Immuni's fault.

**Comment by andreapede** [comment]
Yes, the list in google framework is not correctly sorted but the behaviour
is still the same.
The app now (and like the beginning) communicates only after rebooting the
phone(or rebooting the app).



Il giorno mer 2 set 2020 alle ore 19:00 Riccardo Cossu <
notifications@github.com> ha scritto:

**Comment by Teudimundo** [comment]
I hope this issue is related to this article on "La Repubblica": https://www.repubblica.it/cronaca/2020/09/13/news/bug_di_immuni_mette_a_rischio_il_controllo_dei_contagi-267163423/

If the article is right a fix should be on its way.

**Comment by andreapede** [comment]
It seems that after the lastest update the app started working correctly and polling automatically the server without the need of restarting anything even the software. 
I'll monitor the behavior for  the next days...

**Comment by gahujipo** [comment]
For me it got even worse after the last update. Yesterday I had only one check. 

Xiaomi Mi 6 with Miui Global 11.0.6. All permissions given. All battery savers disabled.

**Comment by andreapede** [comment]
No way. Erratic behavior restarted after 20 days

**Comment by mabete** [comment]
I have the same problem on SAMSUNG SM-N960F, Android 10 Build QP1A.190711.020.N960FXXS6ETHB.
Checking Google/COVID-19 Notification/Exposure Check many days are missing (checked: it's non a problem of day sorting).
If I open Immuni, then many exposure check are logged in the google menu.

**Comment by Teudimundo** [comment]
In my case, since the update of  September 17th, checks are made more often every two/three days, while before they were made only when restarted the phone or reinstalled the app.

**Comment by emavgl** [comment]
The same for me. I have Xioami Redmi Note 8T and the checks are very rare. There are weeks we no check is done, for those days with the checks, I have several checks at the same time (from 2 to 10 checks).

**Comment by simonhunter87** [comment]
Still an issue for me to on a Samsung s20+ in the UK

**Comment by guidodavide** [comment]
Same issue on Samsung S10+ Android 10 QP1A.190711.020.G975FXXU8DTH7
Play Services 20.40.12 (120400-334318821)

My device hasn't updated exposure keys since 15 Sept. I opened Immuni app several times and rebooted the device, enabled/disabled the exposure notification from the Google settings panel didn't help.

After deleting Immuni app and clearing all exposure keys a fresh install was able to download again the keys and sync.
I've lost basically 20 days of encounters though.

Immuni was installed basically since the initial release date, I don't know if this is relevant.

It's unclear whether it's a Google Play Service exposure notification framework issue or Immuni.
I have also a different contact tracing app installed (but not active in the last months) and I tried to enable it once to "force" a key sync. It didn't work.
I suspect the issue is more related to the Google framework, since both applications where not able to update the database after starting them multiple times.
If it's not the framework, it's something related to my Samsung device, which somehow inhibits the key exchange..but I think it's unlikely.
